content/blog/2020-12-16-compcert/index.md

```diff
@@ -0,0 +1,249 @@
+
+#### Measured Forward Simulation
+
+For several of the passes we went over, there might be no corresponding target transition for a source transition, such as a pass that removes `nop` instructions. Particularly, a source program that diverges while a target one gets stuck in one state violates semantic preservation as intuitively defined. To avoid this, we will use a well-founded measure on source states, $<$, to prove that a relation $R_\rightarrow^*$ has the properties:
+
+* $(s_0, t_0) \in R_\rightarrow^*$
+* When $s$ steps to $s'$, one of two things must happen.
+    * $t$ steps to $t'$, and $(s', t') \in R_\rightarrow^*$
+    * $t$ does not step, but $(s', t) \in R_\rightarrow^*$, and $s' < s$.
+
+The measure $<$ allows us to require that the source program progresses as it steps. 
+
+### Composing Simulations
+
+We need to show two things about simulations in order to prove CompCert correct. For the sake of these proof sketches, let's assume that there are no effects or `nop` transitions. 
+
+#### Forward Simulations Compose
+
+Suppose there are two forward simulations: $R_1$ from $\mathbf{S}$ to $\mathbf{M}$, and $R_2$ from $\mathbf{M}$ to $\mathbf{T}$. Let $R_3 = \{(s_i, t_k) | (s_i, m_j) \in R_1 \land (m_j, t_k) \in R_2\}$. We will prove that $R_3$ is a forward simulation from $\mathbf{S}$ to $\mathbf{T}$.
+
+The initial states of $\mathbf{S}$ and $\mathbf{T}$ are related by construction of $R_3$. If $s_i$ steps to $s_{i'}$, and $(s_i, m_j) \in R_1$, we know that there is an $m_{j'}$ such that $m_j\rightarrow m_{j'}$ and $(s_{i'}, m_{j'}) \in R_1$. Similarly, if there is a $t_{k}$ such that $(m_j, t_{k}) \in R_2$, there exists a $t_{k'}$ such that $t_k\rightarrow t_{k'}$ and $(m_{j'}, t_{k'}) \in R_2$. Again by construction, $(s_i, t_k) \in R_3$. Therefore, $t_{k'}$ matches the transition from $s_i$ to $s_{i'}$.
+
+#### Forward Simulations $\rightarrow$ Backward Simulations
+
+The theorem assumes a deterministic target language. We will prove that a forward simulation $R_\rightarrow$ from $\mathbf{S}$ to $\mathbf{T}$ implies a backward simulation $R_\leftarrow$ from $\mathbf{T}$ to $\mathbf{S}$.
+
+Suppose $(s_i, t_j) \in R_\rightarrow$, with $t_j$ stepping to $t_{j'}$. All transitions out of $s_i$ must therefore match the transition from $t_k$, since $t_{j'}$ is unique due to the target language's deterministic semantics. Were this to not hold, the assumed forward simulation would be contradicted.
+
+## Proving Compiler Correctness
+
+Let $\mathcal{C}$ be a compiler, and $\sim$ a simulation as above. To be correct, $\mathcal{C}$ must promise $\mathcal{C}(s) = t \rightarrow s \sim t$ for all source programs $s$. Note that a compilation failure means that this theorem doesn't hold, since there's no $t$. 
```

Comment:
It's not that the _theorem_ doesn't hold (which makes sound like "the theorem is false" which seems very bad)—it's that the _premise_ of some implication doesn't hold, right? So the theorem is still correct, vacuously?

Reply:
Changed "hold" to "apply"; not sure if it's technically vacuously correct since there is no t such that C(s) = t? Vacuousness also confuses me so maybe I'm wrong.